### PR TITLE
fix: Dropdown should respect trigger's disabled prop

### PR DESCRIPTION
**Pull request by afc163**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution

图片

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Fix Button `disabled` prop not working when it is Dropdown's children.         |
| 🇨🇳 Chinese |  修复 Button 作为 Dropdown trigger 时 `disabled` 属性不生效的问题。         |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/dropdown/dropdown.tsx [hunk 1]
```diff
@@ -177,16 +177,19 @@ const Dropdown: CompoundedComponent = (props) => {
...
       `${prefixCls}-trigger`,
       {
         [`${prefixCls}-rtl`]: direction === 'rtl',
       },
       child.props.className,
     ),
-    disabled,
-  });
+  };
+  if (typeof disabled !== 'undefined') {
```
Comment (MadCcc) [inline components/dropdown/dropdown.tsx:189]: 感觉直接合并进去 `disabled: child.props.disabled ?? disabled` 更好呢

File: components/dropdown/dropdown.tsx [hunk 2]
```diff
@@ -177,16 +177,19 @@ const Dropdown: CompoundedComponent = (props) => {
...
+    dropdownTriggerProps.disabled = disabled;
```
Comment (zombieJ) [inline components/dropdown/dropdown.tsx:190]: HOC 和 context 都不会生效:

```tsx
<ConfigProvider componentDisabled>
  <Dropdown><Button /></Dropdown>
</ConfigProvider>
```

直接到 feature 里把这个 disabled 传导去了要不？

File: components/dropdown/dropdown.tsx (hunk 1 above)
Comment (afc163) [reply components/dropdown/dropdown.tsx:189]: 不行，带着 `disabled: undefined` 就会覆盖原来的属性。

Comment (MadCcc) [reply components/dropdown/dropdown.tsx:189]: `??` 会把 `undefined` 过滤掉的

File: components/dropdown/dropdown.tsx (hunk 2 above)
Comment (afc163) [reply components/dropdown/dropdown.tsx:190]: 图片

这个代码实测是生效的。

File: components/dropdown/dropdown.tsx (hunk 1 above)
Comment (afc163) [reply components/dropdown/dropdown.tsx]: 确实可以👍🏻

**Review by li-jia-nan — DISMISSED** [review]

File: components/dropdown/dropdown.tsx [hunk 3]
```diff
@@ -185,7 +185,7 @@ const Dropdown: CompoundedComponent = (props) => {
       },
       child.props.className,
     ),
-    disabled,
+    disabled: disabled ?? child.props.disabled,
```
Comment (li-jia-nan) [inline components/dropdown/dropdown.tsx:188]: 是不是可以换一下顺序？感觉 child 的 disabled 优先级更高一点？

Comment (afc163) [reply components/dropdown/dropdown.tsx]: 确实，已修正，并补充了更多用例。

**Review by li-jia-nan — DISMISSED** [review]

**Review by crazyair — DISMISSED** [review]

File: components/dropdown/dropdown.tsx (hunk 2 above)
Comment (zombieJ) [reply components/dropdown/dropdown.tsx]: 要设置一下 Dropdown 的 `disabled`:

```tsx
<ConfigProvider componentDisabled>
  <Dropdown menu={{ items }} disabled={false}>
    <Button />
  </Dropdown>
</ConfigProvider>
```

截屏2024-02-07 23 53 32

**Review by kiner-tang — DISMISSED** [review]

File: components/dropdown/dropdown.tsx (hunk 2 above)
Comment (afc163) [reply components/dropdown/dropdown.tsx]: 我感觉这样是对的。

File: components/dropdown/dropdown.tsx [hunk 4]
```diff
@@ -176,7 +176,6 @@ const Dropdown: CompoundedComponent = (props) => {
   const [, token] = useToken();
 
   const child = React.Children.only(children) as React.ReactElement<any>;
```
Comment (afc163) [inline components/dropdown/dropdown.tsx:178]:
```suggestion
  const child = React.Children.only(children) as React.ReactElement<any>;

```

**Review by crazyair — APPROVED** [review]

**Pull request merged**